components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,189 @@
+    pub fn exists<T>(&self, ident: &T, target: &PackageTarget) -> Result<bool>
+    where
+        T: Identifiable,
+    {
+        let mut request = ListObjectsRequest::default();
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        request.bucket = self.bucket.to_owned();
+        match self.client.list_buckets().sync() {
+            Ok(bucket_list) => {
+                match bucket_list.buckets {
+                    Some(bkts) => {
+                        for bucket in bkts {
+                            if bucket.name.unwrap() == request.bucket {
+                                match self.client.list_objects(&request).sync() {
+                                    Ok(object_list) => {
+                                        match object_list.contents{
+                                            Some(contents) => {
+                                                for object in contents {
+                                                    if object.key.unwrap() == key {
+                                                        return Ok(true);
+                                                    }
+                                                }
+                                                return Ok(false);
+                                            },
+                                            None => return Ok(false),
+                                        }
+                                    },
+                                    Err(e) => {
+                                        debug!("{:?}", e);
+                                        return Err(Error::ObjectError(e))
+                                    },
+                                }
+                            }
+                        }
+                        return Ok(false);
+                    }
+                    None => Ok(false),
+                }
+            },
+            Err(e) => {
+                debug!("{:?}", e);
+                Err(Error::BadBucket(e))
+            }
+        }
+    }
+
+    pub fn upload(&self, hart: &PathBuf, ident: &PackageIdent, target: &PackageTarget) -> Result<()>
+    {
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        let buck = self.bucket.to_string();
+        let file = File::open(hart).unwrap();
+        let mut reader = BufReader::new(file);
+        let mut object:Vec<u8> = Vec::new();
+        let _result = reader.read_to_end(&mut object);
+
+        let mut request = PutObjectRequest::default();
+        request.key = key;
+        request.bucket = buck;
+        request.body = Some(object);
+        request.content_type = Some("application/x-tar".to_string());
+
+        // TODO (ian): A maximum single PUT to S3 is 5gb
+        // rusoto supports PUT in chunks.  This upload
+        // should chunk the upload request to stay <limit
+        match self.client.put_object(&request).sync() {
```

Comment:
Definitely a concern that we won't have multipart, and we are doing a sync upload.. we should deeply instrument the S3 code with metrics and logging so we have a pretty good understanding of what the perf looks like under different loads.

Reply:
We definitely _will_ have multi-part upload. It just wasn't implemented when I pushed the branch up so I wanted to make sure I had my note in here to get this switched over :)

Reply:
Latest commit implements multi-part uploads, still a raw commit, I didn't get as far as I would have liked but all the functionality should now be in for this the code itself just needs to be cleaned up and potentially optimized.